WT-11303 Move the debug mode retention flags behind a lock

ajmorton opened the pull request:
Take the existing uses of `debug_log_cnt` and `debug_ckpt_cnt` (`debug_mode.log_retention` and `debug_mode.checkpoint_retention` in the connection config) and change them from using memory barriers to locks.

Since this is the first PR in PM-3221 which migrates a barrier to using locks these a few questions to discuss. I'll 
flag them with comments.

The git diff for this PR is a bit hard to read. The main changes are as follows:
- Add a new lock (exact name tbd) which protects reads and writes to `debug_ckpt`, `debug_ckpt_alloc`, `debug_ckpt_cnt`, `debug_log_cnt`
- Pull the logic that reads and writes these fields into subfunctions where needed. This makes it easier to correctly set and release the new lock.

ajmorton commented on src/conn/conn_api.c:
```diff
@@ -2042,11 +2042,7 @@ __wt_debug_mode_config(WT_SESSION_IMPL *session, const char *cfg[])
         FLD_SET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN);
```
This function `__wt_debug_mode_config` is accessed in two places: `__wt_conn_reconfig` which holds `conn->reconfig_lock`, and `wiredtiger_open` which doesn't hold the lock.
Two points to discuss:
- Should we look at taking the `reconfig_lock` during `wiredtiger_open`? We know for a fact we're the only thread at this point but it'd help with correctness checks if we always hold the lock when changing `debug_log_cnt`
- How can we know a lock is held when we enter a function? Some functions have a "Must be called with the XXX lock held" comment but is this enough?

luke-pearson replied:
To quickly answer your second questions see @pmacko86's work on spinlock assertions: https://github.com/wiredtiger/wiredtiger/pull/9049/files.

ajmorton commented on src/conn/conn_log.c:
```diff
@@ -400,12 +401,21 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+     * Take a consistent view of the current configuration. If another thread reconfigures the
+     * connection in parallel we'll handle that change on the next call to this function.
+     */
+    __wt_spin_lock(session, &conn->reconfig_lock);
```
Should we change this to a rwlock? We've identified 16 unique shared variables that on the connection, so our PM-3221 changes could result in multiple threads waiting to hold the `reconfig_lock`. Being able to take read locks here will reduce contention.

Prior to this PR `reconfig_lock` was only used to prevent two concurrent calls to `__wt_conn_reconfig`. I don't see a change from a spinlock to a write lock having much impact.

luke-pearson replied:
I'm in favour of an rwlock.

ajmorton commented on src/conn/conn_log.c:
```diff
@@ -400,12 +401,21 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
-    WT_ORDERED_READ(dbg_val, conn->debug_log_cnt);
-    if (dbg_val != 0) {
+    if (dbg_ckpt_retain && dbg_ckpt_cnt != 0)
+        min_lognum = WT_MIN(conn->debug_ckpt[dbg_ckpt_cnt - 1].l.file, min_lognum);
```
This access to conn->debug_ckpt is outside of the lock but we know it's safe to access it like this (per [this comment](https://github.com/wiredtiger/wiredtiger/blob/fe22c23199fd0f6462a148c51b0a9694857a19bc/src/conn/conn_api.c#L2032) we'll never have a non-zero dbg_ckpt_cnt and an unallocated conn_debug_ckpt array).
Do we play it safe and wrap the whole section in the lock, or try and minimise the time spent holding the lock?

luke-pearson replied:
Play it safe and lock the whole section, conn_reconfig should pay the price of waiting.

luke-pearson commented:
Can you add a comment to `connection.h` indicating that these vars need to be read and written to within the reconfig lock?

ajmorton commented:
We were running into deadlock issues when using `conn->reconfig_lock` as our configuration logic first takes the `reconfig_lock` and then the `checkpoint_lock`, while the checkpoint thread takes the `checkpoint_lock` and with our new code changes would need to take the `reconfig_lock` in `__wt_log_ckpt`.

Instead we've dropped use of the existing reconfig_lock and introduced a new lock for these variables.

ajmorton commented:
@luke-pearson all your comments with a 👍 have been addressed in the latest changes

ajmorton commented on src/include/connection.h:
```diff
@@ -248,17 +248,18 @@ struct __wt_connection_impl {
-    WT_RWLOCK dhandle_lock;      /* Data handle list lock */
+    WT_SPINLOCK api_lock;               /* Connection API spinlock */
+    WT_SPINLOCK checkpoint_lock;        /* Checkpoint spinlock */
+    WT_RWLOCK debug_log_retention_lock; /* Log retention reconfiguration lock */
```
As always naming suggestions are welcome.
The variables being protected are debug variables that control how much of the log we retain. We're specifically protecting them from access while they're in the process of being reconfigured.

luke-pearson replied:
Perhaps `debug_ckpt_log_lock`? It isn't obvious to me that this lock given its current name would apply to debug_ckpt.

ajmorton replied:
In that case I'd prefer to change the naming of `debug_ckpt`. `debug_ckpt` just tracks the LSNs associated with the last N checkpoints, and we use this to determine the oldest log file we're not allowed to discard. Focusing on the checkpoint here feels like a distraction.

luke-pearson replied:
We probably need another persons opinion here who is more familiar with these variables.

luke-pearson commented on src/conn/conn_log.c:
```diff
@@ -400,12 +394,17 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+    __wt_readlock(session, &conn->debug_log_retention_lock);
+
     /* Adjust the number of log files to retain based on debugging options. */
-    WT_ORDERED_READ(dbg_val, conn->debug_ckpt_cnt);
```
👏

luke-pearson commented on src/conn/conn_log.c:
```diff
@@ -400,12 +394,17 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
-    WT_ORDERED_READ(dbg_val, conn->debug_ckpt_cnt);
-    if (FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN) && dbg_val != 0)
-        min_lognum = WT_MIN(conn->debug_ckpt[dbg_val - 1].l.file, min_lognum);
-    WT_ORDERED_READ(dbg_val, conn->debug_log_cnt);
```
👏

luke-pearson commented on src/conn/conn_api.c:
```diff
@@ -2013,40 +2013,61 @@ __wt_extra_diagnostics_config(WT_SESSION_IMPL *session, const char *cfg[])
-     * We need to make sure all writes to other fields are visible before setting the count because
-     * the log removal thread may walk the array using this value.
-     */
-    WT_PUBLISH(conn->debug_ckpt_cnt, (uint32_t)cval.val);
```
👏

luke-pearson commented on src/conn/conn_log.c:
```diff
@@ -400,12 +394,17 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
-    WT_ORDERED_READ(dbg_val, conn->debug_log_cnt);
-    if (dbg_val != 0) {
+
+    dbg_ckpt_cnt = conn->debug_ckpt_cnt;
```
Do we even need dbg_ckpt_cnt here now that we're in a lock? We're not interested in READ_ONCE semantics.

ajmorton replied:
Updated in [ad69de5](https://github.com/wiredtiger/wiredtiger/pull/9428/commits/ad69de5c0b8587bece045368322a982d3de4c89a)

luke-pearson commented on src/conn/conn_log.c:
```diff
@@ -400,12 +394,17 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+
+    dbg_ckpt_cnt = conn->debug_ckpt_cnt;
+    dbg_ckpt_retain = FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN);
+    if (dbg_ckpt_retain && dbg_ckpt_cnt != 0)
```
We could also remove `debug_ckpt_retain`.

ajmorton replied:
Updated in [ad69de5](https://github.com/wiredtiger/wiredtiger/pull/9428/commits/ad69de5c0b8587bece045368322a982d3de4c89a)

luke-pearson commented on src/conn/conn_log.c:
```diff
@@ -400,12 +394,17 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+    if (dbg_ckpt_retain && dbg_ckpt_cnt != 0)
+        min_lognum = WT_MIN(conn->debug_ckpt[dbg_ckpt_cnt - 1].l.file, min_lognum);
+
+    dbg_log_cnt = conn->debug_log_cnt;
```
Another unneeded var.

ajmorton replied:
Updated in [ad69de5](https://github.com/wiredtiger/wiredtiger/pull/9428/commits/ad69de5c0b8587bece045368322a982d3de4c89a)

luke-pearson commented on src/conn/conn_log.c:
```diff
@@ -415,13 +414,42 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+    logfiles = NULL;
+
+    min_lognum = __compute_min_lognum(session, log, backup_file);
+    if (min_lognum == 0)
```
maybe `if (__compute_min_lognum()` dropping the var.

luke-pearson commented on src/include/connection.h:
```diff
@@ -573,6 +574,7 @@ struct __wt_connection_impl {
     int page_size; /* OS page size for mmap alignment */
 
+    /* Access to these fields is protected by the debug_log_retention_lock. */
     WT_LSN *debug_ckpt;      /* Debug mode checkpoint LSNs. */
```
It might be worth doing a CTRL+F for any of these variables to see if they're used outside of locks.

ajmorton replied:
Confirmed. The only use outside of a lock is freeing of `debug_ckpt` in `__wt_connection_destroy`

luke-pearson replied:
Heh, does that need a lock?

keitharnoldsmith reviewed (commented):
LGTM. I left a minor comment here, and a bit of clarification on the ticket.

ajmorton commented on src/conn/conn_api.c:
```diff
@@ -2013,40 +2013,61 @@ __wt_extra_diagnostics_config(WT_SESSION_IMPL *session, const char *cfg[])
- *     Set debugging configuration.
+ * __debug_mode_log_retention_config --
+ *     Set the log retention fields of the debugging configuration. These fields are protected by a
+ *     lock.
```
Thanks! Fixed in [00f427a](https://github.com/wiredtiger/wiredtiger/pull/9428/commits/00f427aa2fe466cf837c20f4f5cabbf44aa04ae3)

luke-pearson reviewed (approved):
LGTM.

ajmorton commented:
evergreen merge
